### PR TITLE
Notify the appropriate team when a scheduled deploy fails 

### PR DESCRIPTION
**Pull request by AWare**
Currently if a scheduled deploy fails, nothing happens. 

This change fires an [a'h'd](https://github.com/guardian/anghammarad) notification when a scheduled deploy fails. 

Before deploying:

- [x] Update Cloudformation
- [x] Add scheduledDeployment.anghammaradTopicARN to PROD configuration

### CODE REVIEW

**Comment by AWare** [comment]
This needs `scheduledDeployment.anghammaradTopicARN` set

File: riff-raff/app/schedule/DeployJob.scala [hunk 1]
```diff
@@ -26,18 +26,22 @@ class DeployJob extends Job with Logging {
       uuid <- deployments.deploy(params, ScheduleRequestSource)
     } yield uuid
     result match {
-      case Left(error) => log.warn(error.message)
+      case Left(error) => {
+        //Send ang msg
```
Comment (alexduf) [inline riff-raff/app/schedule/DeployJob.scala]: Is that comment still relevant?

File: riff-raff/app/conf/context.scala [hunk 2]
```diff
@@ -95,6 +96,13 @@ class Configuration(val application: String, val webappConfDirectory: String = "
 
   object scheduledDeployment {
     lazy val enabled = configuration.getStringProperty("scheduledDeployment.enabled", "false") == "true"
+    lazy val regionName = configuration.getStringProperty("scheduledDeployment.aws.region", "eu-west-1")
+    lazy val snsClient = AmazonSNSAsyncClientBuilder.standard()
+      .withCredentials(credentialsProviderChain(None, None))
+      .withRegion(regionName)
+      .withClientConfiguration(new ClientConfiguration())
```
Comment (regiskuckaertz) [inline riff-raff/app/conf/context.scala]: Is this even necessary?

File: riff-raff/app/notification/ScheduledDeployFailureNotifications.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  def url(uuid: UUID): String = {
+    val path = routes.DeployController.viewUUID(uuid.toString,true)
+    prefix + path.url
+  }
+
+  def failedDeployNotification(uuid: UUID, parameters: DeployParameters) = {
+    val deriveAnghammaradTargets = for {
+      yaml <- TargetResolver.fetchYaml(parameters.build)
+      deployGraph <- Resolver.resolveDeploymentGraph(yaml, deploymentTypes, magenta.input.All).toEither
+      magentaTargets = TargetResolver.extractTargets(deployGraph)
```
Comment (regiskuckaertz) [inline riff-raff/app/notification/ScheduledDeployFailureNotifications.scala]: Save an extra `map` by moving this into the `yield` expression

File: riff-raff/app/schedule/ScheduledDeployer.scala [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package schedule
+
+import magenta.Deployer
+
+object ScheduledDeployer {
```
Comment (regiskuckaertz) [inline riff-raff/app/schedule/ScheduledDeployer.scala]: I don't see where the scheduling happens 😅

File: riff-raff/app/schedule/DeployJob.scala (hunk 1 above)
Comment (jacobwinch) [reply riff-raff/app/schedule/DeployJob.scala]: I think this is intended to capture the case where a scheduled deploy doesn't start at all, which is tracked by a separate card on our board: https://trello.com/c/3jb3GJvl/113-scheduled-deploys-dont-notify-when-they-cant-start

I'll clarify the comment.

File: riff-raff/app/schedule/DeployJob.scala [hunk 5]
```diff
@@ -26,18 +26,22 @@ class DeployJob extends Job with Logging {
       uuid <- deployments.deploy(params, ScheduleRequestSource)
     } yield uuid
     result match {
-      case Left(error) => log.warn(error.message)
+      case Left(error) => {
+        // TODO: send an Anghammarad notification to inform a team that their scheduled deploy didn't start
```
[the remark this answers is not in the document]
Comment (philmcmahon) [reply riff-raff/app/schedule/DeployJob.scala]: ah sorry I missed that!

File: magenta-lib/src/main/scala/magenta/logging.scala [hunk 6]
```diff
@@ -88,7 +88,7 @@ object DeployReporter {
     currentReporter.previousReporter
   }
   def failContext(currentReporter: DeployReporter, message: String, reason: Throwable) {
-    send(currentReporter, Fail(message, reason))
+    send(currentReporter, Fail(message, reason, currentReporter.messageContext.parameters))
```
Comment (sihil) [inline magenta-lib/src/main/scala/magenta/logging.scala]: These changes are unnecessary (see later)

File: magenta-lib/src/main/scala/magenta/logging.scala [hunk 7]
```diff
@@ -124,7 +124,7 @@ object DeployReporter {
 
   private def failException(reporter: DeployReporter, message: String, e: Option[Throwable] = None): FailException = {
     val exception = e.getOrElse(new RuntimeException(message))
-    send(reporter, Fail(message, exception))
+    send(reporter, Fail(message, exception, reporter.messageContext.parameters))
```
Comment (sihil) [inline magenta-lib/src/main/scala/magenta/logging.scala]: ... not necessary ...

File: magenta-lib/src/main/scala/magenta/logging.scala [hunk 8]
```diff
@@ -165,7 +165,7 @@ case class Info(text: String) extends Message
 case class CommandOutput(text: String) extends Message
 case class CommandError(text: String) extends Message
 case class Verbose(text: String) extends Message
-case class Fail(text: String, detail: ThrowableDetail) extends Message
+case class Fail(text: String, detail: ThrowableDetail, parameters: DeployParameters) extends Message
```
Comment (sihil) [inline magenta-lib/src/main/scala/magenta/logging.scala]: ... not necessary ...

File: magenta-lib/src/test/scala/magenta/ReportingTest.scala [hunk 9]
```diff
@@ -137,7 +137,7 @@ class ReportingTest extends FlatSpec with Matchers {
   val verbose = Verbose("return value 0")
   val finishDep = FinishContext(deploy)
   val finishInfo = FinishContext(infoMsg)
-  val failMsg = Fail("Failed", new RuntimeException("Failed"))
+  val failMsg = Fail("Failed", new RuntimeException("Failed"), parameters)
```
Comment (sihil) [inline magenta-lib/src/test/scala/magenta/ReportingTest.scala]: ... not necessary ...

File: riff-raff/app/notification/ScheduledDeployFailureNotifications.scala [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+      case Left(_) =>
+        log.error(s"Failed to derive targets required to notify about failed scheduled deploy: ${uuid}")
+    }
+  }
+
+  def scheduledDeploy(deployParameters: DeployParameters): Boolean = deployParameters.deployer == ScheduledDeployer.deployer
+
+  val messageSub = DeployReporter.messages.subscribe(message => {
+    message.stack.top match {
+      case Fail(_, _, parameters) if scheduledDeploy(parameters) =>
```
Comment (sihil) [inline riff-raff/app/notification/ScheduledDeployFailureNotifications.scala]: `parameters` here are available via `message.context.parameters` and don't need to be included in the `Fail` case class.

File: riff-raff/app/persistence/representation.scala [hunk 11]
```diff
@@ -227,7 +227,7 @@ case class WarningDocument(text: String) extends MessageDocument {
   override lazy val dboFields = List("text" -> text)
 }
 case class FailDocument(text: String, detail: ThrowableDetail) extends MessageDocument {
-  def asMessage(params: DeployParameters, originalMessage: Option[Message]) = Fail(text, detail)
+  def asMessage(params: DeployParameters, originalMessage: Option[Message]) = Fail(text, detail, params)
```
Comment (sihil) [inline riff-raff/app/persistence/representation.scala]: ... not necessary ...

File: riff-raff/app/persistence/representation.scala [hunk 12]
```diff
@@ -247,7 +247,7 @@ object MessageDocument {
       case CommandOutput(text) => CommandOutputDocument(text)
       case CommandError(text) => CommandErrorDocument(text)
       case Verbose(text) => VerboseDocument(text)
-      case Fail(text, detail) => FailDocument(text,detail)
+      case Fail(text, detail, _) => FailDocument(text,detail)
```
Comment (sihil) [inline riff-raff/app/persistence/representation.scala]: ... not necessary ...

File: riff-raff/app/notification/ScheduledDeployFailureNotifications.scala [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+      deployGraph <- Resolver.resolveDeploymentGraph(yaml, deploymentTypes, magenta.input.All).toEither
+    } yield {
+      TargetResolver.extractTargets(deployGraph).toList.flatMap { target =>
+        List(App(target.app), Stack(target.stack))
+      } ++ List(Stage(parameters.stage.name))
+    }
+
+    deriveAnghammaradTargets match {
+      case Right(targets) =>
+        val failureMessage = s"Your scheduled deploy failed. This means that your AMI may become out of date. If no successful deploys occur between now and the next scheduled attempt, it will not be retried."
```
Comment (sihil) [inline riff-raff/app/notification/ScheduledDeployFailureNotifications.scala]: Is this the right messaging? Is the only use case for scheduled deploys AMI updates?

File: riff-raff/app/notification/ScheduledDeployFailureNotifications.scala (hunk 10 above)
Comment (jacobwinch) [reply riff-raff/app/notification/ScheduledDeployFailureNotifications.scala]: 🤦‍♂️Thanks for pointing this out!

File: riff-raff/app/notification/ScheduledDeployFailureNotifications.scala (hunk 13 above)
Comment (jacobwinch) [reply riff-raff/app/notification/ScheduledDeployFailureNotifications.scala]: Good point, I've genericised it (and also added some context about the affected project/build/stage).

**Comment by AWare** [comment]
Thanks Jacob :)

**Comment by jacobwinch** [comment]
> Thanks Jacob :)

No problem, you'd already done most of the hard work!

**Pull request merged**